File: backend/models_pkg/device.py
from __future__ import annotations

from enum import Enum

from sqlalchemy import Column as SAColumn
from sqlalchemy import Enum as SAEnum
from sqlalchemy import Index
from sqlmodel import Field, SQLModel


class DeviceType(str, Enum):
    BACKBONE_GATEWAY = "BACKBONE_GATEWAY"
    POP = "POP"
    CORE_SITE = "CORE_SITE"
    CORE_ROUTER = "CORE_ROUTER"
    EDGE_ROUTER = "EDGE_ROUTER"
    AON_SWITCH = "AON_SWITCH"
    OLT = "OLT"
    ONT = "ONT"
    AON_CPE = "AON_CPE"
    SPLITTER = "SPLITTER"
    HOP = "HOP"
    NVT = "NVT"
    ODF = "ODF"
    BUSINESS_ONT = "BUSINESS_ONT"


class Status(str, Enum):
    UP = "UP"
    DOWN = "DOWN"
    DEGRADED = "DEGRADED"
    BLOCKING = "BLOCKING"


class DeviceRole(str, Enum):
    ACTIVE = "active"
    PASSIVE = "passive"
    ALWAYS_ONLINE = "always_online"


class Device(SQLModel, table=True):
    id: str = Field(primary_key=True, index=True)
    name: str = Field(index=True)
    type: DeviceType = Field(
        sa_column=SAColumn(SAEnum(DeviceType, native_enum=False), nullable=False)
    )
    status: Status = Field(
        default=Status.UP, sa_column=SAColumn(SAEnum(Status, native_enum=False), nullable=False)
    )
    # Provisioning & status fields
    provisioned: bool = False
    admin_override_status: Status | None = Field(
        default=None, sa_column=SAColumn(SAEnum(Status, native_enum=False), nullable=True)
    )
    # Optical placeholders
    tx_power_dbm: float | None = None
    sensitivity_min_dbm: float | None = None
    insertion_loss_db: float | None = None
    # Capacity override in Mbps
    capacity: int | None = None

    class SignalStatus(str, Enum):
        OK = "OK"
        WARNING = "WARNING"
        CRITICAL = "CRITICAL"
        NO_SIGNAL = "NO_SIGNAL"

    signal_power_dbm: float | None = None
    signal_margin_db: float | None = None
    attenuation_db: float | None = None  # Total optical path attenuation
    signal_status: Device.SignalStatus | None = Field(
        default=None, sa_column=SAColumn(SAEnum(SignalStatus, native_enum=False), nullable=True)
    )
    parent_container_id: str | None = Field(default=None, foreign_key="device.id")
    slot_id: str | None = Field(default=None)
    tariff_id: int | None = Field(default=None, foreign_key="tariff.id", index=True)
    vrf_id: int | None = Field(default=None, foreign_key="vrf.id", index=True)
    hardware_model_id: int | None = Field(default=None, foreign_key="hardwaremodel.id", index=True)

    __table_args__ = (Index("ix_device_parent_container_id", "parent_container_id"),)

    def derive_role(self) -> DeviceRole:
        # Passive optical elements (inline path components)
        if self.type in {DeviceType.SPLITTER, DeviceType.HOP, DeviceType.NVT, DeviceType.ODF}:
            return DeviceRole.PASSIVE
        # Always-online restricted: backbone gateway + POP/CORE_SITE only
        if self.type in {DeviceType.BACKBONE_GATEWAY, DeviceType.POP, DeviceType.CORE_SITE}:
            return DeviceRole.ALWAYS_ONLINE
        return DeviceRole.ACTIVE

File: engine-go/internal/models/device.go
package models

import "database/sql"

// DeviceType represents the type/category of a network device
type DeviceType string

const (
	DeviceTypeBackboneGateway DeviceType = "BACKBONE_GATEWAY"
	DeviceTypePOP             DeviceType = "POP"
	DeviceTypeCoreSite        DeviceType = "CORE_SITE"
	DeviceTypeCoreRouter      DeviceType = "CORE_ROUTER"
	DeviceTypeEdgeRouter      DeviceType = "EDGE_ROUTER"
	DeviceTypeAONSwitch       DeviceType = "AON_SWITCH"
	DeviceTypeOLT             DeviceType = "OLT"
	DeviceTypeONT             DeviceType = "ONT"
	DeviceTypeAONCPE          DeviceType = "AON_CPE"
	DeviceTypeSplitter        DeviceType = "SPLITTER"
	DeviceTypeHop             DeviceType = "HOP"
	DeviceTypeNVT             DeviceType = "NVT"
	DeviceTypeODF             DeviceType = "ODF"
	DeviceTypeBusinessONT     DeviceType = "BUSINESS_ONT"
)

// Status represents the operational status of a device or link
type Status string

const (
	StatusUP       Status = "UP"
	StatusDOWN     Status = "DOWN"
	StatusDEGRADED Status = "DEGRADED"
	StatusBLOCKING Status = "BLOCKING"
)

// DeviceRole represents the operational role of a device
type DeviceRole string

const (
	DeviceRoleActive       DeviceRole = "active"
	DeviceRolePassive      DeviceRole = "passive"
	DeviceRoleAlwaysOnline DeviceRole = "always_online"
)

// SignalStatus represents the optical signal quality
type SignalStatus string

const (
	SignalStatusOK       SignalStatus = "OK"
	SignalStatusWarning  SignalStatus = "WARNING"
	SignalStatusCritical SignalStatus = "CRITICAL"
	SignalStatusNoSignal SignalStatus = "NO_SIGNAL"
)

// Device represents a network device (router, switch, ONT, etc.)
type Device struct {
	ID                  string          `db:"id"`
	Name                string          `db:"name"`
	Type                DeviceType      `db:"type"`
	Status              Status          `db:"status"`
	Provisioned         bool            `db:"provisioned"`
	AdminOverrideStatus sql.NullString  `db:"admin_override_status"` // Status or NULL
	TxPowerDBM          sql.NullFloat64 `db:"tx_power_dbm"`
	SensitivityMinDBM   sql.NullFloat64 `db:"sensitivity_min_dbm"`
	InsertionLossDB     sql.NullFloat64 `db:"insertion_loss_db"`
	Capacity            sql.NullInt64   `db:"capacity"` // Mbps override
	SignalPowerDBM      sql.NullFloat64 `db:"signal_power_dbm"`
	SignalMarginDB      sql.NullFloat64 `db:"signal_margin_db"`
	SignalStatus        sql.NullString  `db:"signal_status"` // SignalStatus or NULL
	ParentContainerID   sql.NullString  `db:"parent_container_id"`
	SlotID              sql.NullString  `db:"slot_id"`
	TariffID            sql.NullInt64   `db:"tariff_id"`
	VRFID               sql.NullInt64   `db:"vrf_id"`
	HardwareModelID     sql.NullInt64   `db:"hardware_model_id"`
}

// DeriveRole calculates the device role based on device type
func (d *Device) DeriveRole() DeviceRole {
	// Passive optical elements (inline path components)
	switch d.Type {
	case DeviceTypeSplitter, DeviceTypeHop, DeviceTypeNVT, DeviceTypeODF:
		return DeviceRolePassive
	}

	// Always-online restricted: backbone gateway + POP/CORE_SITE only
	switch d.Type {
	case DeviceTypeBackboneGateway, DeviceTypePOP, DeviceTypeCoreSite:
		return DeviceRoleAlwaysOnline
	}

	return DeviceRoleActive
}

// IsLeaf checks if the device is a leaf device (generates traffic)
func (d *Device) IsLeaf() bool {
	switch d.Type {
	case DeviceTypeONT, DeviceTypeBusinessONT, DeviceTypeAONCPE:
		return true
	}
	return false
}

// EffectiveStatus returns the actual status (admin override if set, else status)
func (d *Device) EffectiveStatus() Status {
	if d.AdminOverrideStatus.Valid {
		return Status(d.AdminOverrideStatus.String)
	}
	return d.Status
}
